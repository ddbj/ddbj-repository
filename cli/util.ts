import { colorize } from 'json_colorize/mod.ts';
import { colors } from 'cliffy/ansi/colors.ts';

export function requireLogin() {
  console.error(`First you need to log in; run ${colors.bold('`ddbj-repository auth login`')}.`);

  Deno.exit(1);
}

export async function ensureSuccess(res: Response) {
  if (res.ok) return;

  const type = res.headers.get('content-type');

  if (type?.includes('application/json')) {
    const payload = await res.json();

    colorize(JSON.stringify(payload, null, 2));
  } else {
    console.error(`Error: ${res.status} ${await res.text()}`);
  }

  Deno.exit(1);
}
