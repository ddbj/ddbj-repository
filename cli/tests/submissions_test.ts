import { assertSpyCall, returnsNext, stub } from 'std/testing/mock.ts';
import type { Stub } from 'std/testing/mock.ts';

import { snapshotTest } from 'cliffy/testing/mod.ts';

import mainCommand from '../main_command.ts';
import { runInContext } from './util.ts';

import submissions1 from './fixtures/submissions-1.json' with { type: 'json' };
import submissions2 from './fixtures/submissions-2.json' with { type: 'json' };

await snapshotTest({
  name: 'submission create 1',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify(submissions1[0])),
      ]
    }, async (fetchStub) => {
      await mainCommand.parse(['submission', 'create', '1']);

      assertSpyCall(fetchStub, 0, {
        args: ['http://example.com/api/submissions/', {
          method: 'POST',
          headers: {
            'Authorization': 'Bearer API_KEY',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ validation_id: 1 }),
        }],
      });
    });
  },
});

await snapshotTest({
  name: 'submission list',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify(submissions1), {
          headers: {
            Link: '<http://example.com/api/submissions?page=2>; rel="next"',
          },
        }),

        new Response(JSON.stringify(submissions2)),
      ]
    }, async () => {
      await mainCommand.parse(['submission', 'list']);
    });
  },
});

await snapshotTest({
  name: 'submission show X-1',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify(submissions1[0])),
      ]
    }, async () => {
      await mainCommand.parse(['submission', 'show', 'X-1']);
    });
  },
});

await snapshotTest({
  name: 'submission get-file X-1 path/to/file',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify(submissions1[0])),
        new Response('foo'),
      ]
    }, async (fetchStub) => {
      await mainCommand.parse(['submission', 'get-file', 'X-1', 'path/to/file']);

      assertSpyCall(fetchStub, 1, {
        args: ['http://example.com/api/validations/1/files/path/to/file', {
          headers: { 'Authorization': 'Bearer API_KEY' },
        }],
      });
    });
  },
});
