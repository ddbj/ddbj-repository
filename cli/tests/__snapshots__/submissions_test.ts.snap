export const snapshot = {};

snapshot[`submission create 1 1`] = `
stdout:
'{
  "id": "X-1",
  "created_at": "2024-03-04T12:45:31.801Z",
  "validation": {
    "id": 1,
    "url": "http://localhost:3000/api/validations/1",
    "user": {
      "uid": "keycloak"
    },
    "db": "JVar",
    "created_at": "2024-03-04T12:45:06.262Z",
    "started_at": "2024-03-04T12:45:06.769Z",
    "finished_at": "2024-03-04T12:45:06.829Z",
    "progress": "finished",
    "validity": "valid",
    "objects": [
      {
        "id": "Excel",
        "files": [
          {
            "path": "validations-1.json",
            "url": "http://localhost:3000/api/validations/1/files/validations-1.json"
          }
        ]
      }
    ],
    "results": [
      {
        "object_id": "_base",
        "validity": "valid",
        "details": null,
        "file": null
      },
      {
        "object_id": "Excel",
        "validity": "valid",
        "details": null,
        "file": {
          "path": "validations-1.json",
          "url": "http://localhost:3000/api/validations/1/files/validations-1.json"
        }
      }
    ],
    "submission": {
      "id": "X-1",
      "url": "http://localhost:3000/api/submissions/X-1"
    }
  }
}
'
stderr:
""
`;

snapshot[`submission list 1`] = `
stdout:
"ID  DB   Created            
--  --   -------            
X-1 JVar 2024-03-04 21:45:31
X-2 JVar 2024-03-04 22:07:11
"
stderr:
""
`;

snapshot[`submission show X-1 1`] = `
stdout:
"ID: X-1
URL: http://example.com/web/submissions/X-1
DB: JVar
Created: 2024-03-04 21:45:31
Validation: #1

Objects:
ID    Files             
Excel validations-1.json
"
stderr:
""
`;

snapshot[`submission get-file X-1 path/to/file 1`] = `
stdout:
"foo
"
stderr:
""
`;
